File: peg-solitaire/fast_dfs.py
"""
Optimized depth-first search for peg solitaire.

Optimizations over dfs.py:
  1. JAX JIT-compiled move generation  — eliminates Python-loop overhead on
                                         the hot inner loop
  2. In-place numpy mutation (apply/undo) — avoids array copies
  3. Move ordering                     — most-open moves first so the win
                                         cutoff fires sooner
  4. Quiescence search                 — extends past the depth limit in
                                         narrow positions (moves <= q)

This module is single-threaded. Parallelism, when wanted, is the caller's
responsibility (e.g. labeling many states concurrently with a thread pool).

Public API
----------
fast_dfs(board, max_depth, q) -> (from_pos, over_pos, to_pos) | None
"""
from __future__ import annotations

import jax
import jax.numpy as jnp
import numpy as np

from board import Board


# ── board conversion ──────────────────────────────────────────────────────────

def _to_array(board: Board) -> np.ndarray:
    arr = np.zeros((board.n, board.n), dtype=np.int8)
    for r, c in board.pegs:
        arr[r, c] = 1
    return arr


# ── candidate move tables ─────────────────────────────────────────────────────

def _build_candidates(
    board: Board,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Precompute all geometrically possible (fr, ov, to) triples."""
    frs, ovs, tos = [], [], []
    for r in range(board.n):
        for c in range(board.n):
            if not board._in_bounds(r, c):
                continue
            for dr, dc in board._DIRECTIONS:
                ovr, ovc = r + dr, c + dc
                tor, toc = r + 2 * dr, c + 2 * dc
                if board._in_bounds(ovr, ovc) and board._in_bounds(tor, toc):
                    frs.append((r, c))
                    ovs.append((ovr, ovc))
                    tos.append((tor, toc))
    return (
        np.array(frs, dtype=np.int32),
        np.array(ovs, dtype=np.int32),
        np.array(tos, dtype=np.int32),
    )


# ── JAX JIT-compiled move generation ─────────────────────────────────────────

@jax.jit
def _valid_mask(
    board_arr: jnp.ndarray,
    frs: jnp.ndarray,
    ovs: jnp.ndarray,
    tos: jnp.ndarray,
) -> jnp.ndarray:
    """Boolean mask over candidate moves that are currently legal."""
    return (
        (board_arr[frs[:, 0], frs[:, 1]] == 1)
        & (board_arr[ovs[:, 0], ovs[:, 1]] == 1)
        & (board_arr[tos[:, 0], tos[:, 1]] == 0)
    )


def _get_moves(arr: np.ndarray, frs, ovs, tos) -> list:
    mask = np.asarray(_valid_mask(arr, frs, ovs, tos))
    idx = np.where(mask)[0]
    return [(tuple(frs[i]), tuple(ovs[i]), tuple(tos[i])) for i in idx]


# ── in-place apply / undo ─────────────────────────────────────────────────────

def _apply(arr, fr, ov, to) -> None:
    arr[fr] = 0
    arr[ov] = 0
    arr[to] = 1


def _undo(arr, fr, ov, to) -> None:
    arr[fr] = 1
    arr[ov] = 1
    arr[to] = 0


# ── move ordering ─────────────────────────────────────────────────────────────

def _order_moves(arr, frs, ovs, tos, moves) -> list:
    """Sort moves by successor count descending (most open first)."""
    scored = []
    for fr, ov, to in moves:
        _apply(arr, fr, ov, to)
        n_succ = int(np.sum(_valid_mask(arr, frs, ovs, tos)))
        _undo(arr, fr, ov, to)
        scored.append((n_succ, fr, ov, to))
    scored.sort(reverse=True)
    return [(fr, ov, to) for _, fr, ov, to in scored]


def _limit_breadth(moves, max_breadth) -> list:
    """Keep at most *max_breadth* moves, discarding the rest.

    Assumes *moves* is already sorted most-open first, so this keeps the moves
    with the most successors.
    """
    if max_breadth is None or len(moves) <= max_breadth:
        return moves
    return moves[:max_breadth]


# ── recursive DFS ─────────────────────────────────────────────────────────────

def _dfs(arr, frs, ovs, tos, depth: int, q: int, max_breadth) -> int:
    moves = _get_moves(arr, frs, ovs, tos)
    if not moves:
        return int(arr.sum())
    if depth == 0 and len(moves) > q:
        return int(arr.sum())

    next_depth = depth - 1 if depth > 0 else 0
    best = int(arr.sum())

    moves = _order_moves(arr, frs, ovs, tos, moves)
    moves = _limit_breadth(moves, max_breadth)
    for fr, ov, to in moves:
        _apply(arr, fr, ov, to)
        result = _dfs(arr, frs, ovs, tos, next_depth, q, max_breadth)
        _undo(arr, fr, ov, to)
        if result < best:
            best = result
            if best == 1:
                break

    return best


# ── public entry point ────────────────────────────────────────────────────────

def fast_dfs(
    board: Board,
    max_depth: int,
    q: int = 1,
    max_breadth: int | None = None,
) -> tuple | None:
    """
    Search *board* to *max_depth* plies and return the move that minimises
    the number of pegs remaining.

    Single-threaded: root moves are evaluated sequentially. The win cutoff
    (1 peg) short-circuits the search.

    Args:
        board:       Board to search from (not mutated).
        max_depth:   Maximum number of moves to look ahead.
        q:           Quiescence threshold — at the depth limit, keep searching
                     if available moves <= q (default 1).
        max_breadth: Maximum number of children to expand per node. When a
                     node has more than *max_breadth* legal moves, the
                     *max_breadth* moves with the most successors are kept and
                     the rest are discarded. None (default) expands every child.

    Returns:
        Best (from_pos, over_pos, to_pos) triple, or None if no moves exist.
    """
    if max_depth < 1:
        raise ValueError("max_depth must be at least 1")
    if max_breadth is not None and max_breadth < 1:
        raise ValueError("max_breadth must be at least 1")

    arr = _to_array(board)
    frs, ovs, tos = _build_candidates(board)

    # Warm up JIT.
    _ = _valid_mask(arr, frs, ovs, tos)

    moves = _get_moves(arr, frs, ovs, tos)
    if not moves:
        return None

    moves = _order_moves(arr, frs, ovs, tos, moves)
    moves = _limit_breadth(moves, max_breadth)

    best_score = int(arr.sum()) + 1
    best_move = None
    for fr, ov, to in moves:
        _apply(arr, fr, ov, to)
        score = _dfs(arr, frs, ovs, tos, max_depth - 1, q, max_breadth)
        _undo(arr, fr, ov, to)
        if score < best_score:
            best_score = score
            best_move = (fr, ov, to)
            if best_score == 1:  # optimal — no need to look further
                break

    return best_move

File: peg-solitaire/cpp/fast_dfs.cpp
// Optimized depth-first search for square-board peg solitaire.
//
// This is a C++ port of fast_dfs.py. The Python version leans on JAX to
// vectorise move generation; in C++ the equivalent tight loops over a small
// precomputed candidate table are already fast, so we keep the same algorithm:
//
//   1. Precomputed candidate table  -- all geometrically possible (fr, ov, to)
//                                       triples, tested for legality in place.
//   2. In-place apply / undo         -- one working array, no copies.
//   3. Move ordering                 -- most-open moves first so the win cutoff
//                                       (1 peg) fires sooner.
//   4. Quiescence search             -- extends past the depth limit in narrow
//                                       positions (moves <= q).
//
// Single-threaded: root moves are evaluated sequentially.
#include "fast_dfs.hpp"

#include <algorithm>
#include <stdexcept>

Candidates build_candidates(const Board& board) {
    Candidates c;
    for (int r = 0; r < board.n; ++r) {
        for (int col = 0; col < board.n; ++col) {
            if (!board.in_bounds(r, col)) {
                continue;
            }
            for (const auto& d : SquareBoard::DIRECTIONS) {
                int ovr = r + d[0], ovc = col + d[1];
                int tor = r + 2 * d[0], toc = col + 2 * d[1];
                if (board.in_bounds(ovr, ovc) && board.in_bounds(tor, toc)) {
                    c.fr.push_back(board.idx(r, col));
                    c.ov.push_back(board.idx(ovr, ovc));
                    c.to.push_back(board.idx(tor, toc));
                }
            }
        }
    }
    return c;
}

namespace {

using Array = std::vector<uint8_t>;

// Count pegs remaining. Cells are 0/1/2 (HOLE/PEG/INVALID), so we count cells
// equal to PEG rather than summing -- INVALID corners must not contribute.
int count_pegs(const Array& arr) {
    int s = 0;
    for (uint8_t v : arr) {
        if (v == Board::PEG) {
            ++s;
        }
    }
    return s;
}

std::vector<Move> get_moves(const Array& arr, const Candidates& c) {
    std::vector<Move> moves;
    for (size_t i = 0; i < c.fr.size(); ++i) {
        if (arr[c.fr[i]] == 1 && arr[c.ov[i]] == 1 && arr[c.to[i]] == 0) {
            moves.push_back({c.fr[i], c.ov[i], c.to[i]});
        }
    }
    return moves;
}

int count_valid(const Array& arr, const Candidates& c) {
    int s = 0;
    for (size_t i = 0; i < c.fr.size(); ++i) {
        if (arr[c.fr[i]] == 1 && arr[c.ov[i]] == 1 && arr[c.to[i]] == 0) {
            ++s;
        }
    }
    return s;
}

inline void apply(Array& arr, const Move& m) {
    arr[m.fr] = 0;
    arr[m.ov] = 0;
    arr[m.to] = 1;
}

inline void undo(Array& arr, const Move& m) {
    arr[m.fr] = 1;
    arr[m.ov] = 1;
    arr[m.to] = 0;
}

// Sort moves by successor count descending (most open first).
void order_moves(Array& arr, const Candidates& c, std::vector<Move>& moves) {
    std::vector<std::pair<int, Move>> scored;
    scored.reserve(moves.size());
    for (const Move& m : moves) {
        apply(arr, m);
        scored.push_back({count_valid(arr, c), m});
        undo(arr, m);
    }
    std::stable_sort(scored.begin(), scored.end(),
                     [](const auto& a, const auto& b) { return a.first > b.first; });
    moves.clear();
    for (auto& p : scored) {
        moves.push_back(p.second);
    }
}

// Keep at most max_breadth moves (max_breadth < 0 means keep all). Assumes
// moves is already sorted most-open first, so this keeps the moves with the
// most successors and discards the rest.
void limit_breadth(std::vector<Move>& moves, int max_breadth) {
    if (max_breadth < 0 || static_cast<int>(moves.size()) <= max_breadth) {
        return;
    }
    moves.resize(static_cast<size_t>(max_breadth));
}

int dfs(Array& arr, const Candidates& c, int depth, int q, int max_breadth) {
    std::vector<Move> moves = get_moves(arr, c);
    if (moves.empty()) {
        return count_pegs(arr);
    }
    if (depth == 0 && static_cast<int>(moves.size()) > q) {
        return count_pegs(arr);
    }

    int next_depth = depth > 0 ? depth - 1 : 0;
    int best = count_pegs(arr);

    order_moves(arr, c, moves);
    limit_breadth(moves, max_breadth);
    for (const Move& m : moves) {
        apply(arr, m);
        int result = dfs(arr, c, next_depth, q, max_breadth);
        undo(arr, m);
        if (result < best) {
            best = result;
            if (best == 1) {
                break;
            }
        }
    }
    return best;
}

}  // namespace

std::optional<Move> fast_dfs(const Board& board, int max_depth, int q,
                             int max_breadth) {
    if (max_depth < 1) {
        throw std::invalid_argument("max_depth must be at least 1");
    }
    if (max_breadth >= 0 && max_breadth < 1) {
        throw std::invalid_argument("max_breadth must be at least 1");
    }

    Array arr = board.cells;
    Candidates c = build_candidates(board);

    std::vector<Move> moves = get_moves(arr, c);
    if (moves.empty()) {
        return std::nullopt;
    }

    order_moves(arr, c, moves);
    limit_breadth(moves, max_breadth);

    int best_score = count_pegs(arr) + 1;
    std::optional<Move> best_move;
    for (const Move& m : moves) {
        apply(arr, m);
        int score = dfs(arr, c, max_depth - 1, q, max_breadth);
        undo(arr, m);
        if (score < best_score) {
            best_score = score;
            best_move = m;
            if (best_score == 1) {  // optimal -- no need to look further
                break;
            }
        }
    }
    return best_move;
}
